### ISSUE TITLE
tensorpack/examples/DeepQNetwork/DQNModel.py

### ISSUE THREAD

**Issue by lezhang-thu** [ppwwyyxx/tensorpack#384]
```
        with tf.variable_scope('target'), \
                collection.freeze_collection([tf.GraphKeys.TRAINABLE_VARIABLES]):
            targetQ_predict_value = self._get_DQN_prediction(next_state)    # NxA
```

Why is `collection.freeze_collection` here? I consulted the `freeze_collection`, and it is said as follows:
```
    """
    Args:
        keys(list): list of collection keys to freeze.
    Returns:
        a context where the collections are in the end restored to its initial state.
    """
```

I don't quite understand why it is done here. I am very grateful if you would tell the reasons. Thanks!

**Comment by ppwwyyxx** [comment]
Variables in this scope don't need to be trained. So they shouldn't be added to the TRAINABLE_VARIABLES collection.

**Comment by lezhang-thu** [comment]
Thanks.
So what is done here is that these variables might be trained, but within that context, these variables would finally be restored to its original value. So anyway, implicitly, they are not trained, right?
I still have a doubt. In the code, there are the following:
`best_v = tf.reduce_max(targetQ_predict_value, 1)    # N,`
or
`best_v = tf.reduce_sum(targetQ_predict_value * predict_onehot, 1)`, depending on whether the DQN variant is double or not. But I feel the point locates at the following code:
`target = reward + (1.0 - tf.cast(isOver, tf.float32)) * self.gamma * tf.stop_gradient(best_v)`

We can notice the `tf.stop_gradient` operation. So is it still necessary to have the `collection.freeze_collection` when creating the target network? I mean the variables within the target network are all `stop_gradient` to some extent.
So my new question is whether `collection.freeze_collection` is still necessary.
I am not questioning the correctness of the code. I know it is correct. I just want to make sure I understand the underlying logic. Thanks a lot!

**Comment by ppwwyyxx** [comment]
No. These variables will not be trained. In tensorflow, only variables added to this collection `tf.GraphKeys.TRAINABLE_VARIABLES` will be trained by default.

Your observation is correct. You can remove `freeze_collection` and it will work the same. But tensorpack will print warnings when some "TRAINABLE_VARIABLES" have no gradient, which is usually a bug. freeze_collection avoid this warning message.

**Comment by lezhang-thu** [comment]
Good! Thanks a lot. Your words are very clear.
Thanks!

**Comment by ppwwyyxx** [comment]
To simplify the code I removed the freeze_collection part in 9850edf5342a27f. It also fixed an import bug which caused crashes in evaluation.

**Comment by lezhang-thu** [comment]
You are so helpful. Thanks a lot!
